lib/elixir/lib/module/checker.ex

```diff
@@ -68,42 +68,60 @@ defmodule Module.Checker do
 
   defp check_remote(module, fun, arity, meta, state) do
     cond do
-      not warn_if_undefined?(module, fun, arity, state) ->
-        []
-
       # TODO: In the future we may want to warn for modules defined
       # in the local context
       Keyword.get(meta, :context_module, false) and state.module != module ->
         []
 
-      # TODO: Add no_autoload
-      not Code.ensure_loaded?(module) ->
+      not Code.ensure_loaded?(module) and warn_undefined?(module, fun, arity, state) ->
         warn(meta, state, {:undefined_module, module, fun, arity})
 
-      not function_exported?(module, fun, arity) ->
+      not function_exported?(module, fun, arity) and warn_undefined?(module, fun, arity, state) ->
         exports = exports_for(module)
         warn(meta, state, {:undefined_function, module, fun, arity, exports})
```

Comment:
I know this was implemented in the previous PR, but I think storing the exports in the warning could become problematic. In particular, some modules might have a lot of exports and storing this information repeatedly could take a lot of memory (and since we're doing some append operations, those are not literals). I think it would be much better to look them up lazily just before printing the warning.

Reply:
It might be better to do it lazily but memory should not be an issue for warnings.